kedro-datasets/kedro_datasets/ibis/table_dataset.py

```diff
@@ -0,0 +1,124 @@
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import TYPE_CHECKING, Any, ClassVar
+
+import ibis.expr.types as ir
+from kedro.io import AbstractDataset, DatasetError
+
+if TYPE_CHECKING:
+    from ibis import BaseBackend
+
+
+class TableDataset(AbstractDataset[ir.Table, ir.Table]):
+    """``TableDataset`` loads/saves data from/to Ibis table expressions.
+
+    Example:
+
+    .. code-block:: pycon
+
+        >>> from kedro_datasets.ibis import TableDataset
+
+    """
+
+    DEFAULT_LOAD_ARGS: ClassVar[dict[str, Any]] = {}
+    DEFAULT_SAVE_ARGS: ClassVar[dict[str, Any]] = {
+        "materialized": "view",
+        "overwrite": True,
+    }
+
+    _connections: ClassVar[dict[tuple[tuple[str, str]], BaseBackend]] = {}
+
+    def __init__(  # noqa: PLR0913
+        self,
+        *,
+        filepath: str | None = None,
+        file_format: str | None = None,
+        table_name: str | None = None,
+        connection: dict[str, Any] | None = None,
+        load_args: dict[str, Any] | None = None,
+        save_args: dict[str, Any] | None = None,
+    ) -> None:
+        """Creates a new ``TableDataset`` pointing to a table (or file).
+
+        Args:
+            filepath: Path to a file to register as a table. Most useful
+                for loading data into your data warehouse (for testing).
+            file_format: Specifies the input file format for `filepath`.
+            table_name: The name of the table or view to read or create.
+            connection: Configuration for connecting to an Ibis backend.
+            load_args: Additional arguments passed to the Ibis backend's
+                `read_{file_format}` method.
+            save_args: Additional arguments passed to the Ibis backend's
+                `create_{materialized}` method. By default, ``ir.Table``
+                objects are materialized as views. To save a table using
+                a different materialization strategy, supply a value for
+                `materialized` in `save_args`.
+        """
+        if filepath is None and table_name is None:
+            raise DatasetError(
+                "Must provide at least one of `filepath` or `table_name`."
+            )
+
+        self._filepath = filepath
+        self._file_format = file_format
+        self._table_name = table_name
+        self._connection_config = connection
+
+        # Set load and save arguments, overwriting defaults if provided.
+        self._load_args = deepcopy(self.DEFAULT_LOAD_ARGS)
+        if load_args is not None:
+            self._load_args.update(load_args)
+
+        self._save_args = deepcopy(self.DEFAULT_SAVE_ARGS)
+        if save_args is not None:
+            self._save_args.update(save_args)
+
+        self._materialized = self._save_args.pop("materialized")
+
+    @property
+    def connection(self) -> BaseBackend:
+        cls = type(self)
+        key = tuple(sorted(self._connection_config.items()))
```

Comment:
I get an error `unhashable type: 'dict'` when using this connection method

```python
conn_table = TableDataset(
    connection={
        "resource": "mssql://xxx.sql.azuresynapse.net/xxx",
        "query": {"driver": "ODBC Driver 17 for SQL Server"},
    },
    table_name="SIGNAL",
    load_args={"schema": "xxx"},
)
```

Reply:
@inigohidalgo Good catch! I've done some digging, and opted to use `json.dumps(self._connection_config, sort_keys=True)` instead. I think this should handle the case you brought up.

Reply:
Ugh, that doesn't work, for non- serialization keys...

Reply:
OK, fixed and added some test cases!